wdl-format/src/format.rs

```diff
@@ -0,0 +1,616 @@
+//! A module for formatting WDL code.
+
+use std::fmt::Write;
+
+use anyhow::Result;
+use format_state::SPACE;
+use wdl_ast::v1::Decl;
+use wdl_ast::v1::DocumentItem;
+use wdl_ast::v1::Expr;
+use wdl_ast::v1::HintsItem;
+use wdl_ast::v1::HintsSection;
+use wdl_ast::v1::InputSection;
+use wdl_ast::v1::LiteralBoolean;
+use wdl_ast::v1::LiteralFloat;
+use wdl_ast::v1::LiteralInteger;
+use wdl_ast::v1::LiteralString;
+use wdl_ast::v1::OutputSection;
+use wdl_ast::v1::StringPart;
+use wdl_ast::v1::StructDefinition;
+use wdl_ast::v1::Type;
+use wdl_ast::AstNode;
+use wdl_ast::AstToken;
+use wdl_ast::Comment;
+use wdl_ast::Diagnostic;
+use wdl_ast::Direction;
+use wdl_ast::Document;
+use wdl_ast::Ident;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+use wdl_ast::Validator;
+use wdl_ast::Version;
+use wdl_ast::VersionStatement;
+
+mod comments;
+mod format_state;
+mod import;
+mod metadata;
+mod task;
+mod workflow;
+
+use comments::format_inline_comment;
+use comments::format_preceding_comments;
+use format_state::FormatState;
+
+/// Newline constant used for formatting.
+pub const NEWLINE: &str = "\n";
+
+/// A trait for elements that can be formatted.
+pub trait Formattable {
+    /// Format the element and write it to the buffer.
+    fn format(&self, buffer: &mut String, state: &mut FormatState) -> Result<()>;
+    /// Get the syntax element of the element.
+    fn syntax_element(&self) -> SyntaxElement;
```

Comment:
That's a good point. There are essentially two possible implementations, one for `Node` and one for `Token`. I should be able to specify the `Node` approach as the default and override as needed? I'll look into it.

Reply:
I was actually going to suggest generalizing, something like:
```
fn syntax_element(&self) -> SyntaxElement {
    match self {
         SyntaxNode => { SyntaxElement::Node(self)}
         SyntaxToken => { SyntaxElement::Token(self)}
    }
}
```

Reply:
That code doesn't actually work. I'm not sure how (or if it's possible) to get something like that working. I assume you'd have to somehow add a trait bound on `self` to indicate that it needs to implement something (`AstNode` or `AstToken`?). More of a Clay/Peter question than something that needs to change here.

Reply:
Interesting discussion, I didn't really like the inclusion of this method in `Formattable`, so it's good we're open to removing the need to implement it for every type (a default implementation or otherwise).

In this case, `rowan` provides a `From` implementation for both [`SyntaxNode`](https://github.com/rust-analyzer/rowan/blob/master/src/api.rs#L85) and [`SyntaxToken`](https://github.com/rust-analyzer/rowan/blob/master/src/api.rs#L91), so there isn't a need for a trait method at all—just import `wdl_grammar::SyntaxElement` and do `SyntaxElement::from(node_or_token)`.

Reply:
> Interesting discussion, I didn't really like the inclusion of this method in `Formattable`, so it's good we're open to removing the need to implement it for every type (a default implementation or otherwise).
> 
> In this case, `rowan` provides a `From` implementation for both [`SyntaxNode`](https://github.com/rust-analyzer/rowan/blob/master/src/api.rs#L85) and [`SyntaxToken`](https://github.com/rust-analyzer/rowan/blob/master/src/api.rs#L91), so there isn't a need for a trait method at all—just import `wdl_grammar::SyntaxElement` and do `SyntaxElement::from(_)`.

That sounds like the right approach to me, @claymcleod . Thanks for pointing this out!